swift/common/middleware/s3api/controllers/cors.py

```diff
@@ -0,0 +1,211 @@
+# Copyright (c) 2018-2020 OpenStack Foundation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from swift.common.utils import public
+
+from swift.common.middleware.s3api.controllers.base import Controller, \
+    bucket_operation, check_container_existence, log_s3api_command
+from swift.common.middleware.s3api.etree import fromstring, \
+    DocumentInvalid, XMLSyntaxError
+from swift.common.middleware.s3api.s3response import HTTPOk, HTTPNoContent, \
+    MalformedXML, NoSuchCORSConfiguration, CORSInvalidRequest
+
+from swift.common.middleware.s3api.utils import sysmeta_header
+
+VERSION_ID_HEADER = 'X-Object-Sysmeta-Version-Id'
+
+MAX_CORS_BODY_SIZE = 10240
+
+BUCKET_CORS_HEADER = sysmeta_header('bucket', 'cors')
+
+CORS_ALLOWED_HTTP_METHOD = ('GET', 'POST', 'PUT', 'HEAD', 'DELETE')
+
+
+def match(pattern, value):
+    '''helper function for wildcard'''
+    if '*' not in pattern:
+        return pattern == value
+    # protect dot as we keep them as is
+    pattern = pattern.replace('.', '\\.')
+    pattern = '^' + pattern.replace('*', '.*') + '$'
+    return re.match(pattern, value) is not None
+
+
+def get_cors(app, req, method, origin):
+    sysmeta = req.get_container_info(app).get('sysmeta', {})
+    body = sysmeta.get('s3api-cors')
+    if not body:
+        return None
+    data = fromstring(body, "CorsConfiguration")
+
+    # we have to iterate over each to find matching origin
+    # whe have to manage wildcard in domain
+    rules = data.findall('CORSRule')
+    for rule in rules:
+        item = rule.find('AllowedOrigin')
+        if match(item.text, origin) or item.text == '*':
+            # check AllowedMethod
+            methods = rule.findall('AllowedMethod')
+            for m in methods:
+                if m.text == method:
+                    hdrs = req.headers.get('Access-Control-Request-Headers')
+                    if hdrs:
+                        allowed = [x.text.lower()
+                                   for x in rule.findall('AllowedHeader')]
+
+                        # manage * as well for headers
+                        hdrs = [x.lower().strip() for x in hdrs.split(',')]
+                        if '*' not in allowed \
+                                and not all([hdr in allowed for hdr in hdrs]):
+                            # some requested headers are not found
+                            continue
+                    return rule
```

Comment:
Each `AllowedHeader` string in the rule can contain at most one * wildcard character. (https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html#cors-allowed-headers)
Also, if headers are not allowed, there is no point in continuing to check with other methods, the result will be the same.
```suggestion
            # check AllowedMethod
            methods = rule.findall('AllowedMethod')
            for method in methods:
                if method.text != method:
                    continue

                headers = req.headers.get('Access-Control-Request-Headers')
                if not headers:
                    return rule

                # check AllowedHeader
                headers = [x.lower().strip() for x in headers.split(',')]
                allowed_headers = [x.text.lower()
                                   for x in rule.findall('AllowedHeader')]
                for header in headers:
                    header_match = False
                    for allowed_header in allowed_headers:
                        if match(allowed_header, header):
                            header_match = True
                            break
                    if not header_match:
                        break
                else:
                    # all requested headers are allowed
                    return rule
                # some requested headers are not allowed
                break
```